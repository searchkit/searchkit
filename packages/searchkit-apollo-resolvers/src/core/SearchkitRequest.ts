import dataloader from 'dataloader'
import { Client, ClientOptions } from '@elastic/elasticsearch'
import HttpAgent, { HttpsAgent } from 'agentkeepalive'
import { SearchkitConfig } from '../resolvers'
import QueryManager from './QueryManager'
import { filterTransform } from './FacetsFns'

export interface SearchResponse<T> {
  took: number
  timed_out: boolean
  hits: {
    total: {
      value: number
    }
    max_score: number
    hits: Array<{
      _id: number
      _source: any
      fields?: any
      highlight?: any
      inner_hits?: any
      matched_queries?: string[]
      sort?: string[]
    }>
  }
  aggregations?: any
}

export const mergeESQueries = (queries) =>
  queries.reduce(
    (combinedQuery, partial) => ({
      ...combinedQuery,
      ...partial,
      aggs: {
        ...combinedQuery.aggs,
        ...partial.aggs
      }
    }),
    {}
  )

const keepaliveHttpsAgent = new HttpsAgent()
const keepaliveAgent = new HttpAgent()

export default class SearchkitRequest {
  private dataloader: any
  private client: Client

  constructor(private queryManager: QueryManager, private config: SearchkitConfig) {
    this.client = new Client({
      node: this.config.host,
      agent: () =>
        new URL(this.config.host).protocol === 'http:' ? keepaliveAgent : keepaliveHttpsAgent
    })

    this.dataloader = new dataloader(async (partialQueries) => {
      const baseQuery = {
        size: 0,
        ...(this.queryManager.hasQuery() && this.config.query
          ? { query: this.config.query.getFilter(this.queryManager) }
          : {}),
        post_filter: filterTransform(this.queryManager, this.config.facets)
      }

      const ESQuery = mergeESQueries([baseQuery, ...(partialQueries as any[])])
      const response = await this.executeQuery(ESQuery)

      return partialQueries.map(() => response)
    })
  }

  public async search(partialQuery): Promise<SearchResponse<any>> {
    return this.dataloader.load(partialQuery)
  }

  private async executeQuery(esQuery): Promise<SearchResponse<any>> {
    try {
      const response = await this.client.search<SearchResponse<any>>({
        index: this.config.index,
        body: esQuery
      })

      return response.body
    } catch (e) {
      console.log(e)
    }
  }
}
