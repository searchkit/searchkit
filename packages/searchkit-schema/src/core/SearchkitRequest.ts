import dataloader from 'dataloader'
import { Client } from '@elastic/elasticsearch'
import HttpAgent, { HttpsAgent } from 'agentkeepalive'
import { SearchkitConfig } from '../resolvers'
import ESQueryError from '../utils/ESQueryError'
import { BaseFacet } from '../facets'
import QueryManager from './QueryManager'
import { filterTransform } from './FacetsFns'

export interface SearchResponse<T> {
  took: number
  timed_out: boolean
  hits: {
    total: {
      value: number
    }
    max_score: number
    hits: Array<{
      _id: number
      _source: any
      fields?: any
      highlight?: any
      inner_hits?: any
      matched_queries?: string[]
      sort?: string[]
    }>
  }
  aggregations?: any
}

export type Query = {
  bool?: {
    must?: Array<Record<string, unknown>>
    should?: Array<Record<string, unknown>>
    filter?: Array<Record<string, unknown>>
  }
}

type BaseQuery = {
  size: number
  query?: Query
  post_filter?: Query
}

export const mergeESQueries = (queries) =>
  queries.reduce(
    (combinedQuery, partial) => ({
      ...combinedQuery,
      ...partial,
      aggs: {
        ...combinedQuery.aggs,
        ...partial.aggs
      }
    }),
    {}
  )

const keepaliveHttpsAgent = new HttpsAgent()
const keepaliveAgent = new HttpAgent()

export default class SearchkitRequest {
  private dataloader: any
  private client: Client

  constructor(
    private queryManager: QueryManager,
    private config: SearchkitConfig,
    private baseFilters: Array<Record<string, unknown>>,
    private facets: Array<BaseFacet>
  ) {
    this.client = new Client({
      node: this.config.host,
      agent: () =>
        new URL(this.config.host).protocol === 'http:' ? keepaliveAgent : keepaliveHttpsAgent
    })

    this.dataloader = new dataloader(async (partialQueries) => {
      const query = this.buildQuery(partialQueries)
      const esQuery = this.config.postProcessRequest ? this.config.postProcessRequest(query) : query
      const response = await this.executeQuery(esQuery)
      return partialQueries.map(() => response)
    })
  }

  public async search(partialQuery): Promise<SearchResponse<any>> {
    return this.dataloader.load(partialQuery)
  }

  private buildQuery(partialQueries): BaseQuery {
    const queryFilter =
      this.queryManager.hasQuery() && this.config.query
        ? this.config.query.getFilter(this.queryManager)
        : null

    const baseFiltersQuery = filterTransform(this.queryManager, this.config.filters)
    const combinedBaseFilters = [].concat(this.baseFilters, baseFiltersQuery?.bool?.must || [])
    const query: Query = queryFilter || (combinedBaseFilters.length > 0 ? {} : null)

    if (combinedBaseFilters.length) {
      if (query.bool) {
        Object.assign(query.bool, {
          filter: query.bool.filter?.length
            ? [].concat(query.bool.filter, combinedBaseFilters)
            : combinedBaseFilters
        })
      } else {
        Object.assign(query, { bool: { filter: combinedBaseFilters } })
      }
    }

    const postFilter = filterTransform(this.queryManager, this.facets)

    let highlight
    this.config.hits.highlightedFields?.forEach((field) => {
      if (!highlight) {
        highlight = { fields: {} }
      }
      if (typeof field == 'string') {
        highlight.fields[field] = {}
      } else {
        highlight.fields[field.field] = field.config
      }
    })

    const baseQuery = { size: 0 }

    return mergeESQueries(
      [
        baseQuery,
        query && { query },
        postFilter && { post_filter: postFilter },
        highlight && { highlight },
        ...(partialQueries as any[])
      ].filter(Boolean)
    )
  }

  private async executeQuery(esQuery): Promise<SearchResponse<any>> {
    const debugMode = process.env.DEBUG_MODE === "true"
    try {
      if (debugMode) console.log(JSON.stringify(esQuery, null, 2))
      const response = await this.client.search<SearchResponse<any>>({
        index: this.config.index,
        body: esQuery
      })

      return response.body
    } catch (e) {
      if (debugMode) console.log(JSON.stringify(e, null, 2))

      if (e.meta?.statusCode === 400) {
        throw new ESQueryError(
          `Elasticsearch query failed. Check your custom filters or configuration. Below is the ES Query`,
          esQuery
        )
      } else {
        throw e
      }
    }
  }
}
