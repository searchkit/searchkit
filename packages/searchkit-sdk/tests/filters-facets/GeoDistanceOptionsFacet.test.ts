import nock from 'nock'
import SearchkitRequest, { SearchkitConfig, GeoDistanceOptionsFacet } from '../../src'
import ResultsNoHitsMock from '../__mock-data__/Facets/results-geo-distance-options.json'

describe('GeoDistanceOptionsFacet', () => {
  it('range of value, min-max range facets', async () => {
    const moviesSearchConfig: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'us_parks',
      hits: {
        fields: []
      },
      facets: [
        new GeoDistanceOptionsFacet({
          field: 'location',
          origin: '37.7749, -122.4194',
          unit: 'mi',
          ranges: [
            { to: 1000, label: '0 - 1000' },
            { to: 2000, label: '0 - 2000' },
            { from: 1000, to: 2000, label: '1000 - 2000' }
          ],
          label: 'location',
          identifier: 'location'
        })
      ]
    }

    const request = SearchkitRequest(moviesSearchConfig)
    request.setFilters([
      {
        identifier: 'location',
        value: '0 - 2000'
      }
    ])

    let lastESRequest

    const scope = nock('http://localhost:9200')
      .post('/us_parks/_search')
      .reply(200, (uri, body: any) => {
        expect(body.aggs).toMatchInlineSnapshot(`
          Object {
            "facet_bucket_all": Object {
              "aggs": Object {
                "location": Object {
                  "geo_distance": Object {
                    "field": "location",
                    "keyed": true,
                    "origin": "37.7749, -122.4194",
                    "ranges": Array [
                      Object {
                        "key": "0 - 1000",
                        "to": 1000,
                      },
                      Object {
                        "key": "0 - 2000",
                        "to": 2000,
                      },
                      Object {
                        "from": 1000,
                        "key": "1000 - 2000",
                        "to": 2000,
                      },
                    ],
                    "unit": "mi",
                  },
                },
              },
              "filter": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "bool": Object {
                        "must": Array [
                          Object {
                            "bool": Object {
                              "must": Array [
                                Object {
                                  "geo_distance": Object {
                                    "distance": "2000mi",
                                    "location": "37.7749, -122.4194",
                                  },
                                },
                              ],
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
              },
            },
          }
        `)
        lastESRequest = body
        return ResultsNoHitsMock
      })
      .persist()

    const response = await request.execute({
      facets: true
    })
    expect(response).toMatchSnapshot()
    expect(response.summary.appliedFilters).toMatchInlineSnapshot(`
      Array [
        Object {
          "display": "ListFacet",
          "id": "location_0 - 2000",
          "identifier": "location",
          "label": "location",
          "type": "GeoDistanceOptionsFacet",
          "value": "0 - 2000",
        },
      ]
    `)
  })
})
