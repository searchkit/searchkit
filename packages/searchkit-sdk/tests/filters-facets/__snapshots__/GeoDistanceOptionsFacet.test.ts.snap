// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeoDistanceOptionsFacet range of value, min-max range facets 2`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 29,
          "label": "0 - 1000",
        },
        Object {
          "count": 42,
          "label": "0 - 2000",
        },
        Object {
          "count": 13,
          "label": "1000 - 2000",
        },
      ],
      "identifier": "location",
      "label": "location",
      "type": "GeoDistanceOptionsFacet",
    },
  ],
  "hits": Object {
    "items": Array [],
    "page": Object {
      "from": 0,
      "pageNumber": NaN,
      "size": 0,
      "total": 42,
      "totalPages": Infinity,
    },
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [
      Object {
        "display": "ListFacet",
        "id": "location_0 - 2000",
        "identifier": "location",
        "label": "location",
        "type": "GeoDistanceOptionsFacet",
        "value": "0 - 2000",
      },
    ],
    "disabledFilters": Array [],
    "query": "",
    "sortOptions": Array [],
    "total": 42,
  },
}
`;
